File: WebClient SDK/src/default/workflow/krProcess/krGlobalTileCommand.ts
import { IKrTileCommand } from './interfaces';
import { Guid } from 'tessa/platform';
import { IUIContext, LoadingOverlay, showConfirm, showNotEmpty } from 'tessa/ui';
import {
  KrSecondaryProcessTileUIHandlerRegistry,
  launchProcess
} from 'tessa/ui/workflow/krProcess';
import { ITile } from 'tessa/ui/tiles';
import { KrProcessInstance, KrTileInfo } from '@tessa/platform';
import { IStorage } from '@tessa/core';

export class KrGlobalTileCommand implements IKrTileCommand {
  //#region ctor

  private constructor() {}

  //#endregion

  //#region instance

  private static _instance: KrGlobalTileCommand;

  public static get instance(): KrGlobalTileCommand {
    if (!KrGlobalTileCommand._instance) {
      KrGlobalTileCommand._instance = new KrGlobalTileCommand();
    }
    return KrGlobalTileCommand._instance;
  }

  //#endregion

  //#region IKrTileCommand members

  public async onClickAction(
    context: IUIContext,
    _tile: ITile,
    tileInfo: KrTileInfo
  ): Promise<void> {
    if (tileInfo.id === Guid.empty) {
      return;
    }

    let additionalStorage: IStorage | undefined = undefined;
    if (tileInfo.handlerId) {
      additionalStorage = {};
      const handler = KrSecondaryProcessTileUIHandlerRegistry.instance.get(tileInfo.handlerId);
      if (handler && !(await handler.handleTile(context, tileInfo, additionalStorage))) {
        return;
      }
    }

    if (tileInfo.askConfirmation) {
      const result = await showConfirm(tileInfo.confirmationMessage);
      if (!result) {
        return;
      }
    }

    await LoadingOverlay.instance.show(async () => {
      const process = KrProcessInstance.create({
        processId: tileInfo.id,
        processInfo: additionalStorage
      });
      const result = await launchProcess(process, { raiseErrorWhenExecutionIsForbidden: true });
      if (result) {
        await showNotEmpty(result.validationResult.build());
      }
    });
  }

  //#endregion
}

File: Extensions/Tessa.Extensions.Default.Client/Workflow/KrProcess/KrGlobalTileCommand.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Localization;
using Tessa.UI;
using Tessa.UI.Tiles;

namespace Tessa.Extensions.Default.Client.Workflow.KrProcess
{
    /// <summary>
    /// Команда выполняющаяся при клике по глобальному тайлу подсистемы маршрутов.
    /// </summary>
    /// <param name="launcher"><inheritdoc cref="IKrProcessLauncher" path="/summary"/></param>
    /// <param name="krSecondaryProcessTileUIHandlerResolver"><inheritdoc cref="IKrSecondaryProcessTileUIHandlerResolver" path="/summary"/></param>
    public sealed class KrGlobalTileCommand(
        IKrProcessLauncher launcher,
        IKrSecondaryProcessTileUIHandlerResolver krSecondaryProcessTileUIHandlerResolver) :
        IKrTileCommand
    {
        #region Fields

        private readonly IKrProcessLauncher launcher = NotNullOrThrow(launcher);
        private readonly IKrSecondaryProcessTileUIHandlerResolver krSecondaryProcessTileUIHandlerResolver = NotNullOrThrow(krSecondaryProcessTileUIHandlerResolver);

        #endregion

        #region IKrTileCommand Members

        /// <inheritdoc />
        public async Task OnClickAsync(
            IUIContext context,
            ITile tile,
            KrTileInfo tileInfo)
        {
            if (tileInfo.ID == Guid.Empty)
            {
                return;
            }

            Dictionary<string, object?>? additionalInfo = null;

            if (tileInfo.HandlerID is not null
                && this.krSecondaryProcessTileUIHandlerResolver.TryResolve(tileInfo.HandlerID.Value) is { } handler
                && !await handler.HandleTileAsync(
                        context,
                        tileInfo,
                        additionalInfo = [],
                        CancellationToken.None))
            {
                return;
            }

            if (tileInfo.AskConfirmation
                && !TessaDialog.Confirm(LocalizeFormat(tileInfo.ConfirmationMessage)))
            {
                return;
            }

            using (TessaSplash.Create("$KrButton_DefaultTileSplash"))
            {
                var process = KrProcessBuilder
                    .CreateProcess()
                    .SetProcess(tileInfo.ID)
                    .SetProcessInfo(additionalInfo)
                    .Build();

                var specificParameters = new KrProcessUILauncher.SpecificParameters
                {
                    RaiseErrorWhenExecutionIsForbidden = true
                };

                var result = await this.launcher.LaunchAsync(
                    process,
                    specificParameters: specificParameters);

                TessaDialog.ShowNotEmpty(result.ValidationResult);
            }
        }

        #endregion
    }
}
